File: app/src/api/star.ts
import axios from "axios";

export interface StarredMessage {
  conversation_id: number;
  message_index: number;
  created_at: string;
}

export async function starMessage(
  conversationId: number,
  messageIndex: number,
): Promise<boolean> {
  try {
    const resp = await axios.post("/conversation/star", {
      conversation_id: conversationId,
      message_index: messageIndex,
    });
    return resp.data.status;
  } catch (e) {
    console.warn(e);
    return false;
  }
}

export async function unstarMessage(
  conversationId: number,
  messageIndex: number,
): Promise<boolean> {
  try {
    const resp = await axios.post("/conversation/unstar", {
      conversation_id: conversationId,
      message_index: messageIndex,
    });
    return resp.data.status;
  } catch (e) {
    console.warn(e);
    return false;
  }
}

export async function getStarredMessages(
  conversationId: number,
): Promise<StarredMessage[]> {
  try {
    const resp = await axios.get(
      `/conversation/starred?conversation_id=${conversationId}`,
    );
    return resp.data.status ? resp.data.data || [] : [];
  } catch (e) {
    console.warn(e);
    return [];
  }
}

File: manager/conversation/star.go
package conversation

import (
	"chat/globals"
	"database/sql"
)

type StarredMessage struct {
	ConversationId int64  `json:"conversation_id"`
	MessageIndex   int    `json:"message_index"`
	CreatedAt      string `json:"created_at"`
}

func StarMessage(db *sql.DB, userId int64, conversationId int64, messageIndex int) error {
	_, err := globals.ExecDb(db, `
		INSERT IGNORE INTO starred_messages (user_id, conversation_id, message_index)
		VALUES (?, ?, ?)
	`, userId, conversationId, messageIndex)
	return err
}

func UnstarMessage(db *sql.DB, userId int64, conversationId int64, messageIndex int) error {
	_, err := globals.ExecDb(db, `
		DELETE FROM starred_messages
		WHERE user_id = ? AND conversation_id = ? AND message_index = ?
	`, userId, conversationId, messageIndex)
	return err
}

func GetStarredMessages(db *sql.DB, userId int64, conversationId int64) []StarredMessage {
	var result []StarredMessage
	rows, err := globals.QueryDb(db, `
		SELECT conversation_id, message_index, created_at
		FROM starred_messages
		WHERE user_id = ? AND conversation_id = ?
		ORDER BY message_index ASC
	`, userId, conversationId)
	if err != nil {
		return result
	}
	defer rows.Close()

	for rows.Next() {
		var item StarredMessage
		if err := rows.Scan(&item.ConversationId, &item.MessageIndex, &item.CreatedAt); err != nil {
			continue
		}
		result = append(result, item)
	}
	return result
}
